### PR TITLE
Implement shapecheck for more primitives

### PR DESCRIPTION
**Pull request by juliuskunze**
**Updated:** All polymorphic shape rules have been removed, and the existing shape rules are reused instead. Surprisingly few and small changes were required on these to generalize. The implementation is powered by overloading the operators necessary for shape manipulation on `Poly`. Allows `shapecheck` of (broadcasting) binops, dilated/strided convs (strides have to divide the strided axis polynomial),  `pad`, `gather`, `broadcast_in_dim`, `device_put`. More `standard_primitive`s might work already, but are not yet tested. Slicing (i. e. `x[1:]`) is not working yet.

This is a step towards my goal of allowing ragged image batches on [PixelCNN++](https://colab.research.google.com/drive/1DMRbUPAxTlk0Awf3D_HR3Oz3P3MBahaJ#scrollTo=uVM7BphAd6Wb). Still missing for this goal are support for slicing + masking rules for all mentioned ops.

### CODE REVIEW

File: jax/interpreters/masking.py [hunk 1]
```diff
@@ -104,10 +104,15 @@ class ShapeExpr(tuple):  # type ShapeExpr = [Poly]
   def __str__(self):
     return 'ShapeExpr({})'.format(', '.join(map(str, self)))
   def __getitem__(self, idx):
-    if type(idx) is int:
-      return super(ShapeExpr, self).__getitem__(idx)
-    else:
-      return ShapeExpr(super(ShapeExpr, self).__getitem__(idx))
+    poly: Poly = super(ShapeExpr, self).__getitem__(idx)
```
Comment (shoyer) [inline jax/interpreters/masking.py]: Just a note: we can't use variable annotations until we drop Python 3.5 support.

Comment (shoyer) [inline jax/interpreters/masking.py]: Just a note: we can't use variable annotations until we drop Python 3.5 support.

Comment (JuliusKunze) [reply jax/interpreters/masking.py]: Good point. They are now removed.

**Comment by JuliusKunze** [comment]
@mattjj Can I remove `ShapeExpr` and replace it with `tuple`? It would make code simpler in many places. Relying on type-checking for `ShapeExpr` seems to be unreliable since indexing into shapes and recombining them through `tuple`s appears to be common. Polymorphic shapes could be consistently identified via any components being of type `Poly`. In fact, this is done already in various places - see usages of `is_polymorphic` [in the diff](https://github.com/google/jax/pull/1990/files).  

Some non-polymorphic shape rules can be implemented using the polymorphic shape rules (see i. e. [conv](https://github.com/google/jax/pull/1990/files#diff-08adae3ce53df47932e06971edb68642R1961) or [pad](https://github.com/google/jax/pull/1990/files#diff-08adae3ce53df47932e06971edb68642R2495) in the diff), avoiding a lot of code duplication. Almost the only difference is the [conversion back from `ShapeExpr` to `tuple`](https://github.com/google/jax/pull/1990/files#diff-08adae3ce53df47932e06971edb68642R1964), which would nicely fall away.

**Comment by mattjj** [comment]
@JuliusKunze removing `ShapeExpr` SGTM. Great points.

You can also write `ShapeExpr = tuple` in the code, i.e. make it a type alias, to express intent in the code that manipulates them. That's just a stylistic suggestion, so if it actually doesn't work out then ignore.

File: jax/interpreters/masking.py [hunk 2]
```diff
@@ -104,30 +106,98 @@ class ShapeExpr(tuple):  # type ShapeExpr = [Poly]
...
-  def __mul__(p1, p2):
-    new_poly = Poly()
-    for (mon1, coeff1), (mon2, coeff2) in it.product(p1.items(), p2.items()):
+  def __add__(self, other):
+    d = self.copy()
+
+    for mon, count in canonicalize_poly(other).items():
+      d[mon] = d.get(mon, 0) + count
+
+    return poly_without_zeros(d)
```
Comment (mattjj) [inline jax/interpreters/masking.py]: Can you say a bit about why you added this method (rather than relying on the one inherited from `Counter`)? (I assume it's to be able to call `poly_without_zeros`.)

File: jax/interpreters/masking.py [hunk 3]
```diff
@@ -104,30 +106,98 @@ class ShapeExpr(tuple):  # type ShapeExpr = [Poly]
...
+  if onp.isscalar(p) and onp.issubdtype(onp.int, onp.array(p).dtype):
+    return constant_poly(p)
+
+  assert isinstance(p, Poly)
+  return p
+
+def canonicalize_polymorphic_shape(shape):
+  return ShapeExpr(map(canonicalize_poly, shape))
+
+def poly_without_zeros(d):
```
Comment (mattjj) [inline jax/interpreters/masking.py]: Can you explain why you need this function? Is it for correctness or just efficiency? A comment could be a good tool here!

File: jax/interpreters/masking.py [hunk 4]
```diff
@@ -104,30 +106,98 @@ class ShapeExpr(tuple):  # type ShapeExpr = [Poly]
...
   def __getitem__(self, idx):
-    if type(idx) is int:
-      return super(ShapeExpr, self).__getitem__(idx)
-    else:
-      return ShapeExpr(super(ShapeExpr, self).__getitem__(idx))
+    poly = super(ShapeExpr, self).__getitem__(idx)
+    return poly if type(idx) is int else ShapeExpr(poly)
+
+
+def canonicalize_poly(p):
```
Comment (mattjj) [inline jax/interpreters/masking.py]: Nit: how about `ensure_poly` as a name rather than `canonicalize_poly`?

File: jax/interpreters/masking.py [hunk 5]
```diff
@@ -104,30 +106,98 @@ class ShapeExpr(tuple):  # type ShapeExpr = [Poly]
...
+  def __divmod__(self, divisor):
+    if self.is_constant:
+      q, r = divmod(int(self), divisor)
+
+      return constant_poly(q), r
+
+    def divided(count):
+      q, r = divmod(count, divisor)
+      if r != 0:
+        raise ValueError('shapecheck currently only supports strides '
```
Comment (mattjj) [inline jax/interpreters/masking.py]: I think we can support division when there's also an added constant, i.e. `c * poly + k` can be safely integer-divided by `c`. If that's right, can you add a note or a todo here about being able to support that case? If not, teach me why not :)

File: jax/lax/lax.py [hunk 6]
```diff
@@ -4256,7 +4219,22 @@ def _dilate_shape(shape, dilation):
   return onp.where(shape == 0, 0,
                    onp.multiply(dilation, onp.subtract(shape, 1)) + 1)
 
+def is_shape_nonnegative(shape):
```
Comment (mattjj) [inline jax/lax/lax.py]: Could you prefix the name of this function and the other similar utility functions with an underscore?

**Review by mattjj — APPROVED** [review]
It's extremely impressive how you just dove into this and improved it substantially. Awesome.

Only some minor suggestions and requests for explanation (for my own edification).

Thanks so much for this!

**Comment by mattjj** [comment]
We might get merge conflicts with #1998 but I think just renaming ones.

File: jax/interpreters/masking.py (hunk 2 above)
Comment (JuliusKunze) [reply jax/interpreters/masking.py]: Tried this, but `Counter.__add__` doesn't work correctly with negative values. These are needed to express polynomials like `n-1`.

File: jax/interpreters/masking.py (hunk 5 above)
Comment (JuliusKunze) [reply jax/interpreters/masking.py]: Great point! Will add this.

File: jax/lax/lax.py (hunk 6 above)
Comment (JuliusKunze) [reply jax/lax/lax.py]: Sure.

File: jax/interpreters/masking.py (hunk 2 above)
Comment (mattjj) [reply jax/interpreters/masking.py]: That makes perfect sense, thanks for figuring it out and explaining!

File: jax/interpreters/masking.py (hunk 4 above)
Comment (JuliusKunze) [reply jax/interpreters/masking.py]: Sounds good!

File: jax/interpreters/masking.py (hunk 3 above)
Comment (JuliusKunze) [reply jax/interpreters/masking.py]: This function is now removed.

**Comment by JuliusKunze** [comment]
@mattjj Ended up removing `ShapeExpr` completely since the remaining places where it could have been applied as an alias were quite arbitrary.

File: jax/interpreters/masking.py (hunk 3 above)
Comment (mattjj) [reply jax/interpreters/masking.py]: Actually, I was asking about `_poly_without_zeros`, though I realize now you may have been answering about `canonicalize_polymorphic_shape`.

Why do you need `_poly_without_zeros`? Just pruning / efficiency?

File: jax/interpreters/masking.py (hunk 5 above)
Comment (JuliusKunze) [reply jax/interpreters/masking.py]: This is now implemented.

File: tests/masking_test.py [hunk 7]
```diff
@@ -83,6 +145,91 @@ def cat(x, y, z):
...
+  def test_shapecheck_jit(self):
+    @shapecheck(['n'], '2*n')
+    @jit
+    def concat(x):
+      return lax.concatenate([x, x], 0)
+
+  def test_shapecheck_pad(self):
+    @shapecheck(['n'], '2*n+1')
+    def p(x):
+      return lax.pad(x, 0, [(1, 1, 1)])
```
Comment (mattjj) [inline tests/masking_test.py]: Like Chomsky's poverty of the stimulus, these checks only ensure that correct programs don't cause errors, but they don't check that incorrect programs do cause errors. It's not necessary for this PR, but we should probably add tests for the negative side too.

File: jax/interpreters/masking.py (hunk 3 above)
Comment (JuliusKunze) [reply jax/interpreters/masking.py]: Ah, I see. I was just fixing a bug related to this. I inlined `_poly_without_zeros` into the `Poly` constructor now and added a line to explain the purpose: It's to keep `Poly` in canonical form in order to simplify operations like comparison etc.

Comment (mattjj) [reply jax/interpreters/masking.py]: Excellent, that makes sense! Thanks for explaining.

**Comment by mattjj** [comment]
Ready to merge, or still iterating? (We can land and iterate, of course, but I wanted to check for a green light on your end.)

**Comment by JuliusKunze** [comment]
@mattjj Let's merge, I'll put slicing + masking into a new PR.

**Comment by mattjj** [comment]
Woo awesome! Will merge when Travis passes.

Btw as you probably know, strided slicing is like strided conv, in that it'll require division (and hence engage the same constraints we have for strided conv).

**Comment by JuliusKunze** [comment]
@mattjj Yes. Slicing with polymorphic shape gets currently tangled up [here](https://github.com/google/jax/blob/master/jax/numpy/lax_numpy.py#L3112). Will have a better look tomorrow.
